### PR TITLE
portal: Only derive key if needed - take 2

### PR DESCRIPTION
**Pull request by sophie-h**
Replaces #18

### CODE REVIEW

**Comment by sophie-h** [comment]
I don't think there was a reason to switch to RwLock, to begin with. Mutex is usually faster than RwLock.

**Comment by sophie-h** [comment]
But we can also stick with rwlock

**Comment by sophie-h** [comment]
I will just revert to RwLock and everyone is happy.

**Comment by A6GibKm** [comment]
My reasoning for using RwLock is that the api is nicer and that we do multiple reads but only a few writes. It might be possible that RwLock has too big of an overhead for the small amount of operations that we make use  but I have no clue if thats the case.

**Comment by A6GibKm** [comment]
I don't have a strong opinion on the matter, but  RwLock vs Mutex is outside of the scope of this PR as far as I can tell.

**Comment by bilelmoussaoui** [comment]
> I don't have a strong opinion on the matter, but RwLock vs Mutex is outside of the scope of this PR as far as I can tell.

I agree, it would be faster to review one thing at a time and land it. I would like to do a new beta release to push this crate a bit forward.

**Comment by sophie-h** [comment]
> RwLock vs Mutex is outside of the scope of this PR as far as I can tell.

Well, your change was based on your assumption that it would make the feature in this PR possible. So I reverted it in the implementation that does not need it.

**Comment by sophie-h** [comment]
Ready to go from my side

File: src/portal/secret.rs [hunk 1]
```diff
@@ -10,9 +10,29 @@ use rand::{distributions::Alphanumeric, thread_rng, Rng};
 #[cfg(feature = "tokio")]
 use tokio::{io::AsyncReadExt, net::UnixStream};
 use zbus::zvariant::{Fd, ObjectPath, OwnedValue, SerializeDict, Type};
+use zeroize::{Zeroize, ZeroizeOnDrop};
 
 use super::Error;
 
+#[derive(Debug, Zeroize, ZeroizeOnDrop)]
+pub struct Secret {
```
Comment (bilelmoussaoui) [inline src/portal/secret.rs:18]: As Secret is now part of the public API, maybe it needs to be re-exported in src/portal/mod.rs? or event src/lib.rs?

Comment (bilelmoussaoui) [reply src/portal/secret.rs:18]: Also could use a small comment about what this type represents

File: src/portal/mod.rs [hunk 2]
```diff
@@ -103,48 +106,60 @@ impl Keyring {
...
-            key,
+            key: Default::default(),
+            secret,
         })
     }
 
+    /// Retrieve the number of items
+    ///
+    /// This function will not trigger a key derivation.
+    pub async fn n_items(&self) -> usize {
```
Comment (bilelmoussaoui) [inline src/portal/mod.rs:121]: I would be interested in having this implemented also for the dbus implementation & exposed in the "common" generic API if possible.

File: src/portal/mod.rs [hunk 3]
```diff
@@ -217,18 +236,26 @@ impl Keyring {
     }
 }
 
+/// Return key, derive and store it first if not initialized
+async fn derive_key<'a>(
+    key: &'a mut OnceCell<crate::Key>,
+    keyring: &api::Keyring,
+    secret: &[u8],
```
Comment (bilelmoussaoui) [inline src/portal/mod.rs:243]:
```suggestion
    secret: &Secret,
```
is more accurate no?

File: src/portal/secret.rs (hunk 1 above)
Comment (sophie-h) [reply src/portal/secret.rs:18]: > maybe it needs to be re-exported in src/portal/mod.rs

It is re-exported in `src/portal/mod.rs` already. I added docs to it.

File: src/portal/mod.rs (hunk 3 above)
Comment (sophie-h) [reply src/portal/mod.rs]: changed

File: src/portal/mod.rs (hunk 2 above)
Comment (sophie-h) [reply src/portal/mod.rs:121]: Anyways, this PR only touches the `portal` part anyways so the rest can be done in a different PR.

**Pull request merged**